extensions/ql-vscode/src/debugger/debug-session.ts

```diff
@@ -0,0 +1,587 @@
+  }
+
+  protected quickEvalRequest(
+    response: CodeQLProtocol.QuickEvalResponse,
+    args: CodeQLProtocol.QuickEvalRequest["arguments"],
+  ): void {
+    switch (this.state) {
+      case "stopped":
+        // Send the response immediately. We'll send a "stopped" message when the evaluation is complete.
+        this.sendResponse(response);
+
+        // For built-in requests that are expected to cause execution (`launch`, `continue`, `step`, etc.),
+        // the adapter does not send a `continued` event because the client already knows that's what
+        // is supposed to happen. For a custom request, though, we have to notify the client.
+        this.sendEvent(new ContinuedEvent(QUERY_THREAD_ID, true));
+
+        void this.evaluate(args.quickEvalContext);
+        break;
+
+      default:
+        this.unexpectedState(response);
+        break;
+    }
+  }
+
+  /** Creates a `BaseLogger` that sends output to the debug console. */
+  private createLogger(): BaseLogger {
+    return {
+      log: async (message: string, _options: LogOptions): Promise<void> => {
+        this.sendEvent(new OutputEvent(message, "console"));
+      },
+    };
+  }
+
+  /**
+   * Runs the query or quickeval, and notifies the debugger client when the evaluation completes.
+   *
+   * This function is invoked from the `launch` and `continue` handlers, without awaiting its
+   * result.
+   */
+  private async evaluate(
+    quickEvalContext: QuickEvalContext | undefined,
+  ): Promise<void> {
+    const args = this.args!;
+
+    this.tokenSource = new CancellationTokenSource();
+    try {
+      // Create the query run, which will give us some information about the query even before the
+      // evaluation has completed.
+      this.queryRun = this.queryRunner.createQueryRun(
+        args.database,
+        {
+          queryPath: args.query,
+          quickEvalPosition: quickEvalContext?.quickEvalPosition,
+        },
+        true,
+        args.additionalPacks,
+        args.extensionPacks,
+        this.queryStorageDir,
+        undefined,
+        undefined,
+      );
+
+      this.state = "running";
+
+      // Send the `EvaluationStarted` event first, to let the client known where the outputs are
+      // going to show up.
+      this.sendEvent(
+        new EvaluationStartedEvent(
+          this.queryRun.id,
+          this.queryRun.outputDir.querySaveDir,
+          quickEvalContext,
+        ),
+      );
+
+      try {
+        // Report progress via the debugger protocol.
+        const progressStart = new ProgressStartEvent(
+          this.queryRun.id,
+          "Running query",
+          undefined,
+          0,
+        );
+        progressStart.body.cancellable = true;
+        this.sendEvent(progressStart);
+        let result: CoreCompletedQuery;
+        try {
+          result = await this.queryRun.evaluate(
+            (p) => {
+              const progressUpdate = new ProgressUpdateEvent(
+                this.queryRun!.id,
```

Comment:
To avoid using `!` here, I would propose storing the query run in a local variable as well as storing it in the instance variable.